tiledb/sm/query/readers/aggregators/count_aggregator.h

```diff
@@ -0,0 +1,123 @@
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines class CountAggregator.
+ */
+
+#ifndef TILEDB_COUNT_AGGREGATOR_H
+#define TILEDB_COUNT_AGGREGATOR_H
+
+#include "tiledb/common/status.h"
+#include "tiledb/sm/enums/layout.h"
+#include "tiledb/sm/query/readers/aggregators/iaggregator.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+class QueryBuffer;
+
+class CountAggregator : public IAggregator {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  CountAggregator() = delete;
+
+  /**
+   * Constructor.
+   *
+   * @param field_name Name of the field that this aggregate will process.
+   */
+  CountAggregator(const std::string field_name);
+
+  DISABLE_COPY_AND_COPY_ASSIGN(CountAggregator);
+  DISABLE_MOVE_AND_MOVE_ASSIGN(CountAggregator);
+
+  /* ********************************* */
+  /*                API                */
+  /* ********************************* */
+
+  /** Returns the field name for the aggregator. */
+  std::string field_name() override {
+    return constants::all_attributes;
+  }
+
+  /** Returns if the aggregation is var sized or not. */
+  bool var_sized() override {
+    return false;
+  };
+
+  /** Returns if the aggregate needs to be recomputed on overflow. */
+  bool need_recompute_on_overflow() override {
+    return true;
+  }
+
+  /**
+   * Validate the result buffer.
+   *
+   * @param output_field_name Name for the output buffer.
+   * @param buffers Query buffers.
+   */
+  void validate_output_buffer(
+      std::string output_field_name,
+      std::unordered_map<std::string, QueryBuffer>& buffers) override;
+
+  /**
+   * Aggregate data using the aggregator.
+   *
+   * @param input_data Input data for aggregation.
+   */
+  void aggregate_data(AggregateBuffer& input_data) override;
+
+  /**
+   * Copy final data to the user buffer.
+   *
+   * @param output_field_name Name for the output buffer.
+   * @param buffers Query buffers.
+   */
+  void copy_to_user_buffer(
+      std::string output_field_name,
+      std::unordered_map<std::string, QueryBuffer>& buffers) override;
+
+ private:
+  /* ********************************* */
+  /*         PRIVATE ATTRIBUTES        */
+  /* ********************************* */
+
+  /** Cell count. */
+  std::atomic<uint64_t> count_;
```

Comment:
This is confusing to me? Are we intending to only have one count for all fields? Or is this logic like "we're counting all non-null results" so the only thing that makes sense where all attributes are non-null? But then I'm confused how this count is different than just counting the end result of the query.

Reply:
Granted, I am reading this PR top to bottom, so maybe it'll be clear later.

Reply:
This is counting the end results of the query... Today, in a regular query, we can count the results using the buffers but this will enable a way to quickly get a result count for a query without requesting any data. For example this will evaluate much faster: SELECT COUNT(*) WHERE dim2 > 10 AND dim2 < 20.